vespalib/src/vespa/vespalib/net/tls/crypto_codec.h

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include <memory>
+
+namespace vespalib::net::tls {
+
+struct HandshakeResult {
+    // Handshake bytes consumed from peer.
+    size_t bytes_consumed = 0;
+    // Handshake bytes produced that must be sent to the peer.
+    size_t bytes_produced = 0;
+    enum class State {
+        Failed,
+        Done,
+        NeedsMorePeerData
+        // TODO add Closed/Shutdown as own state?
+    };
+    State state = State::Failed;
+
+    bool failed() const noexcept { return (state == State::Failed); }
+    bool done() const noexcept { return (state == State::Done); }
+};
+
+struct EncodeResult {
+    // Plaintext bytes consumed
+    size_t bytes_consumed = 0;
+    // Ciphertext bytes produced that must be sent to the peer
+    size_t bytes_produced = 0;
+    bool failed = true;
+};
+
+struct DecodeResult {
+    // Ciphertext bytes consumed from peer
+    size_t bytes_consumed = 0;
+    // Plaintext bytes produced.
+    size_t bytes_produced = 0;
+    enum class State {
+        Failed,
+        OK,
+        NeedsMorePeerData
+        // TODO add Closed/Shutdown as own state?
+    };
+    State state = State::Failed;
+
+    bool failed() const noexcept { return (state == State::Failed); }
+};
+
+class TlsContext;
+
+// TODO move to different namespace, not dependent on TLS?
+
+/*
+ * A CryptoCodec provides a fully transport-independent way of negotiating
+ * a secure, authenticated session towards another peer. The codec requires
+ * the caller to handle any and all actual data transfer
+ */
+class CryptoCodec {
+public:
+    enum class Mode {
+        Client, Server
+    };
+
+    virtual ~CryptoCodec() = default;
+
+    virtual size_t min_encode_buffer_size() const noexcept = 0;
+
+    /*
+     * Postcondition: if result.done(), the handshake process has completed
+     *                and data may be passed through encode()/decode().
+     */
+    virtual HandshakeResult handshake(const char* from_peer, size_t from_peer_buf_size,
+                                      char* to_peer, size_t to_peer_buf_size) noexcept = 0;
+
+    /*
+     * Encodes a single ciphertext frame into `ciphertext`. If plaintext_size
+     * is greater than can fit into a frame, the returned result's consumed_bytes
+     * field will be < plaintext_size. The number of actual ciphertext bytes produced
+     * is available in the returned result's produced_bytes field.
+     *
+     * Precondition: handshake must be completed
```

Comment:
There is a precondition listed on `ciphertext_size`, but should we add one on `plaintext_size` as well? Currently you can throw an arbitrarily large `plaintext` buffer (well, < `INT32_MAX`...) at the function and it will only encode max 1 frame and return the number of bytes it actually consumed from the `plaintext` buffer.

Reply:
my bad. should only have precondition on output buffer